### PR TITLE
Add Link model & "create link" endpoint.

### PR DESCRIPTION
**Pull request by DFurnes**
### What's this PR do?

This pull request adds a "create link" endpoint, based on Bertly's [existing `POST /` endpoint](https://github.com/DoSomething/bertly/tree/master/documentation#create-short-link).

### How should this be reviewed?

I've added the [Dynamoose](https://dynamoosejs.com/getting_started/Introduction) (DynamoDB) model in 2919e2f. This sets up our database schema, indexes, and some database settings (could maybe extract if we ever add more models).

I then added the `POST /` endpoint & a functional test in 2371773. The most interesting thing here is how this creates unique short-links and partitions the counter to avoid a single "hot" key.

### Any background context you want to provide?

I still need to set up validation & authentication but will tackle that later on.

I'll push up PRs for the other endpoints once this is merged (since they need `Link` too).

### Relevant tickets

References [Pivotal #172799437](https://www.pivotaltracker.com/story/show/172799437).

### Checklist

- [ ] This PR has been added to the relevant Pivotal card.
- [ ] Added appropriate feature/unit tests.

### CODE REVIEW

File: src/Functions/createLink.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    return res.json(existing);
+  }
+
+  // If we haven't yet shortened this URL, make new shortlink:
+  const key = await generateKey();
+  const link = await Link.create({ key, url });
+
+  info('Created new shortlink', pick(link, ['key', 'url']));
+
+  return res.status(201).json(link);
```
Comment (DFurnes) [inline src/Functions/createLink.js:60]: This won't match the old response format. I'm thinking I'll add a `Transformers/` directory with functions to handle this in a follow-up pull request. Some things are inconsistent (like `url` is the short URL here, and the long URL on the old "clicks" endpoint... going to see if I can unify those without breaking anything).

File: src/Models/Link.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+const options = {
+  // If we're running in development, wait for the local server to be started and
+  // automatically run a "migration" to create DynamoDB table if it doesn't exist.
+  // (We skip this on production instances since it has a performance impact).
+  create: ['development', 'test'].includes(process.env.NODE_ENV),
+  waitForActive: ['development', 'test'].includes(process.env.NODE_ENV),
+
+  // TODO: This seems to be bugged <...>. We should see if we can figure out what's
+  // going wrong here, and fix if possible so we can run "update" migrations.
+  update: false,
```
Comment (DFurnes) [inline src/Models/Link.js:42]: This was driving me mental – the index would exist the first time I ran the application & then as soon as I'd reload the app, everything would explode with an `The table does not have the specified index: urlIndex` error.

After scratching my head for a while (and then adding this handy [query logger](https://github.com/DoSomething/bertly/blob/master/src/Logging/DynamoLogger.js)), I figured out that if we already have this `urlIndex` when the application starts, Dynamoose's "update" migrator will just delete it. And therefore, every other time it runs the application will either add or remove this index:

Screen Shot 2020-05-13 at 10 53 00 AM

To keep things sane, I've disabled this option for now (so we _only_ run migrations the first time the application boots). In the relatively uncommon scenario where we change this schema, we can just delete the local database to re-migrate with a new schema.

I'll revisit this error later on & see if I can track down more details to file an issue in the library.

**Comment by sheyd** [comment]
Every time I read `Dynamoose`, I can't help but think of this. 

image

File: src/Functions/createLink.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * @return {String}
+ */
+async function generateKey() {
+  // In order to avoid a single "hot" key with our counter, we split
+  // the table into 3,844 partitions (e.g. AA, AB, AC...). Each partition
+  // keeps its own auto-incremented counter for making shortlinks, below:
+  const partition = await Link.update(
+    { key: `${randomChar()}${randomChar()}` },
+    { $ADD: { counter: 1 } }
```
Comment (weerd) [inline src/Functions/createLink.js:21]: Is this some specific property key or something? `$ADD`. Just unfamiliar.

File: src/Functions/createLink.js (hunk 1 above)
Comment (weerd) [reply src/Functions/createLink.js:60]: 🤞

File: src/Functions/createLink.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ */
+export default async function createLink(req, res) {
+  // TODO: Validate input and return a 422 for bad stuff.
+  const { url } = req.body;
+
+  // Find & return already shortened link, if one exists.
+  const result = await Link.query({ url }).limit(1).exec();
+  const existingLink = result.count ? result[0] : null;
+  if (existingLink) {
+    info('Found shortlink', pick(existingLink, ['key', 'url']));
```
Comment (weerd) [inline src/Functions/createLink.js:49]: Never used `pick()` before! Pretty nice little Lodash func!

File: src/Models/Link.js (hunk 2 above)
Comment (weerd) [reply src/Models/Link.js:42]: Jeez that sounds like it was painful to debug 😬

File: src/Functions/createLink.js (hunk 3 above)
Comment (DFurnes) [reply src/Functions/createLink.js:21]: Yep! This is a special property on the [`Model.update`](https://dynamoosejs.com/guide/Model#modelupdatekey-updateobj-settings-callback) payload that maps to this [DynamoDB query](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html#Expressions.UpdateExpressions.SET.IncrementAndDecrement). It allows us to increment this counter without a multi-step read & write transaction.

File: src/Models/Link.js (hunk 2 above)
Comment (DFurnes) [reply src/Models/Link.js:42]: It was a spooky one!! 👻

**Pull request merged**